File: conformance/tests/auction/auction.runar.ts
import { StatefulSmartContract, assert, checkSig, extractLocktime } from 'runar-lang';

class Auction extends StatefulSmartContract {
  readonly auctioneer: PubKey;
  highestBidder: PubKey;
  highestBid: bigint;
  readonly deadline: bigint;

  constructor(auctioneer: PubKey, highestBidder: PubKey, highestBid: bigint, deadline: bigint) {
    super(auctioneer, highestBidder, highestBid, deadline);
    this.auctioneer = auctioneer;
    this.highestBidder = highestBidder;
    this.highestBid = highestBid;
    this.deadline = deadline;
  }

  public bid(sig: Sig, bidder: PubKey, bidAmount: bigint) {
    assert(checkSig(sig, bidder));
    assert(bidAmount > this.highestBid);
    assert(extractLocktime(this.txPreimage) < this.deadline);
    this.highestBidder = bidder;
    this.highestBid = bidAmount;
  }

  public close(sig: Sig) {
    assert(checkSig(sig, this.auctioneer));
    assert(extractLocktime(this.txPreimage) >= this.deadline);
  }
}

File: conformance/tests/auction/auction.runar.rs
use runar::prelude::*;

#[runar::contract]
struct Auction {
    #[readonly]
    auctioneer: PubKey,
    highest_bidder: PubKey,
    highest_bid: Bigint,
    #[readonly]
    deadline: Bigint,
}

#[runar::methods(Auction)]
impl Auction {
    #[public]
    fn bid(&mut self, sig: Sig, bidder: PubKey, bid_amount: Bigint) {
        assert!(check_sig(sig, bidder));
        assert!(bid_amount > self.highest_bid);
        assert!(extract_locktime(self.tx_preimage) < self.deadline);
        self.highest_bidder = bidder;
        self.highest_bid = bid_amount;
    }

    #[public]
    fn close(&mut self, sig: Sig) {
        assert!(check_sig(sig, self.auctioneer));
        assert!(extract_locktime(self.tx_preimage) >= self.deadline);
    }
}
